File: vvd-common/src/main/java/mikufan/cx/vvd/common/naming/FileNameUtil.java
package mikufan.cx.vvd.common.naming;

import mikufan.cx.vvd.common.vocadb.model.SongForApi;

/**
 * @author CX无敌
 * @date 2020-12-19
 */
public interface FileNameUtil {

  static String buildErrorInfoJsonFileName(SongForApi song){
    return buildBasicFileNameForSong(song) + FileNamePostFix.SONG_INFO_ERR + ".json";
  }

  static String buildInfoJsonFileName(SongForApi song){
    return buildBasicFileNameForSong(song) + FileNamePostFix.SONG_INFO + ".json";
  }

  static String buildResourceJsonFileName(SongForApi song){
    return buildBasicFileNameForSong(song) + FileNamePostFix.RESOURCES + ".json";
  }

  static String buildPvFileName(SongForApi song, String extensionWithDot){
    return buildBasicFileNameForSong(song) + FileNamePostFix.VIDEO + extensionWithDot;
  }

  static String buildThumbnailFileName(SongForApi song, String extensionWithDot){
    return buildBasicFileNameForSong(song) + FileNamePostFix.THUMBNAIL + extensionWithDot;
  }

  static String buildAudioFileName(SongForApi song, String extensionWithDot){
    return buildBasicFileNameForSong(song) + FileNamePostFix.AUDIO + extensionWithDot;
  }

  static String buildBasicFileNameForSong(SongForApi song){
    var artists = song.getArtistString().split("feat\\.");
    var vocals = artists[1].trim();
    var producers = artists[0].trim();
    var songName = song.getName();
    return removeIllegalChars(String.format("【%s】%s【%s】", vocals, songName, producers));
  }

  static String removeIllegalChars(String fileName){
    return fileName
        .replace("/", "-")
        .replace("\\", "-")

        .replace("? ", " ")
        .replace("* ", " ")
        .replace(": ", " ")

        .replace("?", " ")
        .replace("*", " ")
        .replace(":", " ");
  }
}

File: vvd-common-kt/src/main/kotlin/mikufan/cx/vvd/commonkt/naming/FileNameUtil.kt
package mikufan.cx.vvd.commonkt.naming

import mikufan.cx.vocadbapiclient.model.SongForApiContract
import mikufan.cx.vvd.common.naming.FileNamePostFix
import mikufan.cx.vvd.commonkt.exception.orThrowVocaloidExp

/**
 * @author CX无敌
 * @date 2020-12-19
 */

fun SongForApiContract.toPvFileName(extensionWithDot: String = ""): String {
  return this.toProperFileName() + FileNamePostFix.VIDEO + extensionWithDot
}

fun SongForApiContract.toThumbnailFileName(extensionWithDot: String = ""): String {
  return this.toProperFileName() + FileNamePostFix.THUMBNAIL + extensionWithDot
}

fun SongForApiContract.toAudioFileName(extensionWithDot: String = ""): String {
  return this.toProperFileName() + FileNamePostFix.AUDIO + extensionWithDot
}

// not sure if above extensions are needed

fun SongForApiContract.toErrorFileName(): String {
  return this.toProperFileName() + FileNamePostFix.ERROR + ".json"
}

fun SongForApiContract.toProperFileName(): String {
  val artists: List<String> = artistString.orThrowVocaloidExp("artist string is null").split("feat.")
  val vocals = artists[1].trim()
  val producers = artists[0].trim()
  val songName: String = defaultName.orThrowVocaloidExp("song name is null")
  return removeIllegalChars(String.format("【%s】%s【%s】", vocals, songName, producers))
}

fun removeIllegalChars(fileName: String): String {
  return fileName
    .replace("/", "-")
    .replace("\\", "-")
    .replace("? ", " ")
    .replace("* ", " ")
    .replace(": ", " ")
    .replace("?", " ")
    .replace("*", " ")
    .replace(":", " ")
}
